internals/src/read_uint_iter.rs

```diff
@@ -0,0 +1,36 @@
+// SPDX-License-Identifier: CC0-1.0
+
+/// Reads a `usize` from an iterator of bytes.
+pub fn read_uint_iter(
+    data: &mut core::slice::Iter<'_, u8>,
+    size: usize,
+) -> Result<usize, UintError> {
+    if data.len() < size {
+        Err(UintError::EarlyEndOfScript)
+    } else if size > usize::from(u16::MAX / 8) {
+        // Casting to u32 would overflow
+        Err(UintError::NumericOverflow) // FIXME: Add another error variant for here?
+    } else {
+        let mut ret = 0;
+        for (i, item) in data.take(size).enumerate() {
+            ret = usize::from(*item)
+                // Casting is safe because we checked above to not repeat the same check in a loop
+                .checked_shl((i * 8) as u32)
+                .ok_or(UintError::NumericOverflow)? // FIXME: This is related to `size` not overflow.
```

Comment:
This check looks broken because it would "overflow" also when `size` is larger than `size_of::<usize>()` but there are just zeroes in the data. However it could only ever happen on 16-bit architectures which we explicitly do not support because the largest number we're passing in is `4`.

Reply:
Yeah, I would suggest we just `debug_assert` (or even `assert` that we only pass in 1 through 4).

If it simplified anything I'd also suggest special-casing 1 through 4, though I don't think it does.

Reply:
I was thinking of just having an enum `UintKind { PushData1 = 1, PushData2 = 2, PushData4 = 4 }` and accept that. That gets rid of all ambiguity. We don't even need to derive anything on it.

Reply:
Wow, turns out there is a whole bunch of problems with this function (if I'm not mistaken): #3293